CRM/MembershipExtras/Hook/Alter/CalculatedMembershipStatus.php

```diff
@@ -0,0 +1,336 @@
+<?php
+
+/**
+ * Calculates membership status taking into account in_arrears and not_arrears
+ * start and end events.
+ */
+class CRM_MembershipExtras_Hook_Alter_CalculatedMembershipStatus {
+
+  private $membership;
+  private $calculationArguments;
+  private static $memberShipStatuses = array();
+  private static $contributionStatusValueMap = array();
+
+  public function __construct() {
+    if (count(self::$memberShipStatuses) == 0) {
+      $membershipStatuses = civicrm_api3('MembershipStatus', 'get', [
+        'sequential' => 1,
+        'options' => ['sort' => 'weight ASC', 'limit' => 0],
+      ]);
+      self::$memberShipStatuses = $membershipStatuses['values'];
+    }
+
+    if (count(self::$contributionStatusValueMap) == 0) {
+      $contributionStatuses = civicrm_api3('OptionValue', 'get', array(
+        'option_group_id' => "contribution_status",
+      ));
+
+      foreach ($contributionStatuses['values'] as $currentStatus) {
+        self::$contributionStatusValueMap[$currentStatus['name']] = $currentStatus['value'];
+      }
+    }
+  }
+
+  /**
+   * This hook is called when membership status is being calculated.
+   *
+   * @param array $calculatedStatus
+   *   Membership status details as determined by CiviCRM core
+   * @param array $arguments
+   *   Arguments passed in to calculate date
+   * @param array $membership
+   *   Membership details from the calling function
+   */
+  public function alterMembershipStatus(&$calculatedStatus, $arguments, $membership) {
+    $this->membership = $membership;
+    $this->calculationArguments = $arguments;
+
+    foreach (self::$memberShipStatuses as $currentStatus) {
+      $startEventIsArrearsRelated = stripos($currentStatus['start_event'], 'arrears') !== false;
+      $endEventIsArrearsRelated = stripos($currentStatus['end_event'], 'arrears') !== false;
+
+      if (!$startEventIsArrearsRelated && !$endEventIsArrearsRelated) {
+        // If we reach calculated status, we don't need to consider other options by weight.
+        if ($calculatedStatus['id'] == $currentStatus['id']) {
+          break;
+        }
+
+        // No arrears, so we continue with next status.
+        continue;
+      }
+
+      $startEvent = $this->checkEvent(
+        CRM_Utils_Array::value('start_event', $currentStatus),
+        $arguments['status_date'],
+        CRM_Utils_Array::value('start_event_adjust_unit', $currentStatus),
+        CRM_Utils_Array::value('start_event_adjust_interval', $currentStatus)
+      );
+
+      $endEvent = $this->checkEvent(
+        CRM_Utils_Array::value('end_event', $currentStatus),
+        $arguments['status_date'],
+        CRM_Utils_Array::value('start_event_adjust_unit', $currentStatus),
+        CRM_Utils_Array::value('start_event_adjust_interval', $currentStatus)
+      );
+
+      if ($startEvent && !$endEvent) {

```

Comment:
I don't see the point of !$endEvent here to be honest, at the end it is just the inverse of $startEvent check so I don't think we need it all.